### PR TITLE
Conversion between heterogenous and homogeneous graph objects

### LINKED ISSUE #22: Heterogeneous to typed homogeneous graph conversion

### PR DESCRIPTION
**Pull request by yzhao062**
This PR aims to fix #22. For now, currently the heterogeneous to homogenous graph is implemented. 
In the current implementation, we infer whether all the nodes have the same dimensionalities and combine them if that is True.  If not all the nodes have the same number of dimensions, the node attributes will not be combined. 

The same inference process is also applied to the edge attributes.

`# call conversion
data_homogeneous = data.to_homogeneous()
`

After the conversion, the node and edge slice information is saved in `_node_slice` and `_edge_slice`, which will be used for the conversion back to the heterogeneous graph.

We open this draft for discussion. Pending items include:

- support the specific node attribute combination as a parameter. similar for the edge attributes.
- add the functionality for converting back a heterogeneous graph  

-

### CODE REVIEW

File: test/data/test_hetero_data.py [hunk 1]
```diff
@@ -123,3 +123,35 @@ def test_copy_hetero_data():
...
+    data['paper'].x = torch.randn(100, 128)
+    data['author'].x = torch.randn(200, 128)
+    data['institute'].x = torch.randn(300, 128)
+
+    # add a few edges and attributes
+    data['paper', 'cites', 'paper'].edge_index = \
+        torch.randint(low=0, high=data['paper'].num_nodes, size=[2, 300])
+    data['paper', 'cites', 'paper'].edge_attr = torch.randn(300, 56)
+
+    data['author', 'cites', 'paper'].edge_index = \
```
Comment (rusty1s) [inline test/data/test_hetero_data.py:141]: Not that this will only give you edges pointing to the first 100 authors nodes. I did something similar [here](https://github.com/vrtex-team/pytorch_geometric/blob/master/test/loader/test_hgt_loader.py#L16-L35), which might be helpful nonetheless.

File: test/data/test_hetero_data.py [hunk 2]
```diff
@@ -123,3 +123,35 @@ def test_copy_hetero_data():
...
+        torch.randint(low=0, high=data['paper'].num_nodes, size=[2, 150])
+    data['author', 'cites', 'paper'].edge_attr = torch.randn(150, 56)
+
+    data['institute', 'hires', 'author'].edge_index = \
+        torch.randint(low=0, high=data['author'].num_nodes, size=[2, 180])
+    data['institute', 'hires', 'author'].edge_attr = torch.randn(180, 56)
+
+    data_homogeneous = data.to_homogeneous()
+
+    assert (data_homogeneous.num_nodes == 600)
```
Comment (rusty1s) [inline test/data/test_hetero_data.py:151]: We should also check for node feature equivalence.

File: torch_geometric/data/hetero_data.py [hunk 3]
```diff
@@ -385,3 +387,132 @@ def get_edge_store(self, src: str, rel: str, dst: str) -> EdgeStorage:
...
+            data_homogeneous = data.to_homogeneous()
+
+        """
+
+        # 1. combine all the nodes
+        node_slices = {}
+        node_idx_tracker = 0
+
+        # iterate over all nodes and record the slice information
+        for n, node_store in enumerate(self.node_stores):
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:408]:
```suggestion
        for node_type, node_store in self._node_stores_dict.items():
```

File: torch_geometric/data/hetero_data.py [hunk 4]
```diff
@@ -385,3 +387,132 @@ def get_edge_store(self, src: str, rel: str, dst: str) -> EdgeStorage:
...
+        data_homogeneous = Data(x=node_features,
+                                edge_index=edge_index,
+                                edge_attr=edge_features,
+                                _node_slice=node_slices,
+                                _edge_slice=edge_slice)
+
+        # todo: should we delete the original data object?
+        return data_homogeneous
+
+    def _have_consistent_dims(self, stores):
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:494]: I think this should give you a list of attribute names that have a consistent dimensionality, which is useful for other attributes such as labels `y` or node positions `pos` as well. You can iterate over all possible attributes via
```python
for key, value in store.items():
    pass
```
and infer if dimensions (except the node dimension):
```python
sizes_dict = defaultdict(set)
for key, value in store.items():
    if isinstance(value, Tensor):
       dim = self.__cat_dim__(key, value, store)
       size = tuple(value.size()[:dim] + value.size()[dim + 1:])
       sizes_dict[key].add(size)
        
return [key for key, sizes in sizes_dict.items() if len(sizes) == 1]
```

File: torch_geometric/data/hetero_data.py [hunk 5]
```diff
@@ -385,3 +387,132 @@ def get_edge_store(self, src: str, rel: str, dst: str) -> EdgeStorage:
...
+
+        # decide if node attributes can be combined automatically
+        node_feature_consistent, num_node_features_global = \
+            self._have_consistent_dims(self.node_stores)
+
+        # if the dimensionality matches for all nodes, then combine node
+        # attributes
+        node_features = None
+
+        if node_feature_consistent:
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:422]: Why not make use of `torch.cat`?
```python
dim = self.__cat_dim__(...)  
x = torch.cat([store.x for store in self.node_stores], dim=dim)
```

File: torch_geometric/data/hetero_data.py [hunk 6]
```diff
@@ -385,3 +387,132 @@ def get_edge_store(self, src: str, rel: str, dst: str) -> EdgeStorage:
...
+            node_features = torch.zeros(
+                [self.num_nodes, num_node_features_global])
+            for node_key in node_slices.keys():
+                # combine node attributes
+                node_features[
+                    node_slices[node_key][0]:node_slices[node_key][1], :] \
+                    = self[node_key]['x']
+
+        # 2. combine all the edges
+        edge_index = torch.zeros([2, self.num_edges], dtype=torch.long)
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:433]: `torch.cat`

File: torch_geometric/data/hetero_data.py [hunk 7]
```diff
@@ -385,3 +387,132 @@ def get_edge_store(self, src: str, rel: str, dst: str) -> EdgeStorage:
...
+            # get the type of current edge_store
+            edge_type = self.edge_types[e]
+
+            # given the different node type, we need to update the edge_index
+            # for the new node index
+            edge_index_offset = torch.zeros_like(edge_store['edge_index'],
+                                                 dtype=torch.long)
+
+            # for source node, add the offset
+            edge_index_offset[0, :] = node_slices[edge_type[0]][0]
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:451]: Make use of broadcasting:
```python
edge_index = edge_index + torch.tensor([[node_slices...], [node_slices...]]
```

File: torch_geometric/data/hetero_data.py [hunk 8]
```diff
@@ -385,3 +387,132 @@ def get_edge_store(self, src: str, rel: str, dst: str) -> EdgeStorage:
...
+            for edge_key in edge_slice.keys():
+                # combine edge attributes
+                edge_features[edge_slice[
+                                  edge_key][0]:edge_slice[edge_key][1], :] = \
+                    self[edge_key].edge_attr
+
+        # 3. initialize a `torch_geometric.data.Data` with additional
+        # information to convert the data object back
+
+        data_homogeneous = Data(x=node_features,
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:485]: Instead of `edge_slice`, the conversion should create a new attribute called `edge_type`, holding a vector with values in `{0, ..., num_edge_types - 1}`, i.e., for use in [`RGCNConv`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.RGCNConv).
